api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java

```diff
@@ -0,0 +1,216 @@
+
+import java.util.*;
+
+import static java.util.Locale.ENGLISH;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DiagnosticReportObsTranslatorHelperTest {
+	
+	private static final String REPORT_URL = "/100/uploadReport.pdf";
+	
+	private static final String REPORT_NAME = "bloodTest.pdf";
+	
+	private static final String LAB_TEST_NOTES = "Report is normal";
+	
+	private static final int PATIENT_ID = 12345;
+	
+	@Mock
+	private ConceptService conceptService;
+	
+	@InjectMocks
+	private final DiagnosticReportObsTranslatorHelper diagnosticReportObsResultTranslatorHelper = new DiagnosticReportObsTranslatorHelper();
+	
+	@Test
+    public void shouldTranslateToObsWhenTestReportIsUploadedWithLabNotesForSingleTestWithoutAnyOrder() {
+        Concept testConcept = new Concept();
+        Patient patient = new Patient();
+        patient.setId(PATIENT_ID);
+        Date testDate = new Date();
+        LabResult labResult = new LabResult();
+        labResult.setPatient(patient);
+        labResult.setObservationDate(testDate);
+        labResult.setConcept(testConcept);
+        labResult.setLabReportUrl(REPORT_URL);
+        labResult.setLabReportFileName(REPORT_NAME);
+        labResult.setLabReportNotes(LAB_TEST_NOTES);
+        mockConceptServiceGetConceptByName();
+
+        Obs obsModel = diagnosticReportObsResultTranslatorHelper.createObs(labResult);
+
+        assertPatientAndObservationDate(testDate, obsModel);
+        assertEquals(testConcept, obsModel.getConcept());
+        Set<Obs> obsGroupMembersTopLevel = obsModel.getGroupMembers();
+        assertEquals(1, obsGroupMembersTopLevel.size());
+
+        Obs obsModelSecondLevel = obsGroupMembersTopLevel.iterator().next();
+        assertPatientAndObservationDate(testDate, obsModelSecondLevel);
+        assertEquals(testConcept, obsModelSecondLevel.getConcept());
+        Set<Obs> obsGroupMembersSecondLevel = obsModelSecondLevel.getGroupMembers();
+        assertEquals(3, obsGroupMembersSecondLevel.size());
+
+        obsGroupMembersSecondLevel.forEach(obsModelChildLevel -> {
+            assertPatientAndObservationDate(testDate, obsModelChildLevel);
+            String obsChildConceptName = obsModelChildLevel.getConcept().getPreferredName(ENGLISH).getName();
+            switch (obsChildConceptName) {
+                case "LAB_REPORT":
+                    assertEquals(REPORT_URL, obsModelChildLevel.getValueText());
+                    break;
+                case "LAB_NOTES":
+                    assertEquals(LAB_TEST_NOTES, obsModelChildLevel.getValueText());
+                    break;
+                case "LAB_RESULT":
+                    assertEquals(REPORT_NAME, obsModelChildLevel.getValueText());
+                    break;
+            }
```

Comment:
this test would pass even if all the 3 observations are duplicate

Reply:
Yes. Also, no sort of validation present in the frontend. Do we need that as well?

Reply:
This has been taken up in PR #4